Air-gapped environment

ytimocin opened the pull request:
# Description

This pull request introduces significant enhancements and new features across multiple areas, including Kubernetes deployment, configuration management, and development environment initialization. Key changes include adding support for air-gapped environments, refining Kubernetes installation options, and improving the handling of provisioning states in resources.

### Kubernetes Deployment Enhancements:
* Added the ability to disable the Contour ingress controller during Radius installation (`pkg/cli/cmd/install/kubernetes/kubernetes.go`). [[1]](diffhunk://#diff-969a1495c18d847fc09b31b1e6854e7be42d541513807361aaa5e9497a62e86fR54-R56) [[2]](diffhunk://#diff-969a1495c18d847fc09b31b1e6854e7be42d541513807361aaa5e9497a62e86fR81-R90) [[3]](diffhunk://#diff-969a1495c18d847fc09b31b1e6854e7be42d541513807361aaa5e9497a62e86fR100-R112) [[4]](diffhunk://#diff-969a1495c18d847fc09b31b1e6854e7be42d541513807361aaa5e9497a62e86fR146-R151)
* Updated the `deploy-local.sh` script to automate building, pushing Docker images, resetting the Kubernetes cluster, and publishing Bicep extensions (`deploy-local.sh`).

### Air-Gapped Environment Support:
* Introduced the `--recipe-registry` flag in the `rad init` command to specify an alternative container registry for dev recipes (`pkg/cli/cmd/radinit/init.go`). [[1]](diffhunk://#diff-043b8a9fb42a3b3d2876c07a59c8975147156ed05f3e2c30c934e8a0e65d2ecfR68-R79) [[2]](diffhunk://#diff-043b8a9fb42a3b3d2876c07a59c8975147156ed05f3e2c30c934e8a0e65d2ecfR88) [[3]](diffhunk://#diff-043b8a9fb42a3b3d2876c07a59c8975147156ed05f3e2c30c934e8a0e65d2ecfR121-R123) [[4]](diffhunk://#diff-043b8a9fb42a3b3d2876c07a59c8975147156ed05f3e2c30c934e8a0e65d2ecfR176-R180)
* Modified the `GetDevRecipes` method to accept a `recipeRegistry` parameter for fetching recipes from a custom registry (`pkg/cli/cmd/radinit/mock_devrecipeclient.go`, `pkg/cli/cmd/radinit/environment.go`). [[1]](diffhunk://#diff-dcf64385937a5726c0b18e03a39fde36885ef35051b75518f152a874f13963daL71-R71) [[2]](diffhunk://#diff-2918474decdd8eda8947fb6dad7690cf22e764b0f77e3e2b9a0b4d7a36a06699L44-R55) [[3]](diffhunk://#diff-2918474decdd8eda8947fb6dad7690cf22e764b0f77e3e2b9a0b4d7a36a06699L71-R77)

### Configuration Improvements:
* Added new parameters and resources to the `app-kubernetes-postgres.bicep` file for managing secrets, authentication, and registry settings (`app-kubernetes-postgres.bicep`).
* Updated the `bicepconfig.json` file to use a local extension path instead of a registry reference (`bicepconfig.json`).

### Resource Management:
* Refactored the `ProvisioningState` methods in `BaseResource` to use a simpler structure (`pkg/armrpc/api/v1/types.go`).

### Testing and Mock Updates:
* Adjusted tests and mocks to support the new `recipeRegistry` parameter (`pkg/cli/cmd/radinit/init_test.go`, `pkg/cli/cmd/radinit/mock_devrecipeclient.go`). [[1]](diffhunk://#diff-a6aea1316b5b38221ce3b6e03cb2d3fb22a0f339f4952190d4ca1a6c2614cb5aL820-R820) [[2]](diffhunk://#diff-2918474decdd8eda8947fb6dad7690cf22e764b0f77e3e2b9a0b4d7a36a06699L44-R55) [[3]](diffhunk://#diff-2918474decdd8eda8947fb6dad7690cf22e764b0f77e3e2b9a0b4d7a36a06699L71-R77)

## Type of change
- This pull request adds or changes features of Radius and has an approved issue (issue link required).

Fixes: #issue_number

## Contributor checklist
Please verify that the PR meets the following requirements, where applicable:

- An overview of proposed schema changes is included in a linked GitHub issue.
    - [ ] Yes <!-- TaskRadio schema -->
    - [x] Not applicable <!-- TaskRadio schema -->
- A design document PR is created in the [design-notes repository](https://github.com/radius-project/design-notes/), if new APIs are being introduced.
    - [ ] Yes <!-- TaskRadio design-pr -->
    - [x] Not applicable <!-- TaskRadio design-pr -->
- The design document has been reviewed and approved by Radius maintainers/approvers.
    - [ ] Yes <!-- TaskRadio design-review -->
    - [x] Not applicable <!-- TaskRadio design-review -->
- A PR for the [samples repository](https://github.com/radius-project/samples) is created, if existing samples are affected by the changes in this PR.
    - [ ] Yes <!-- TaskRadio samples-pr -->
    - [x] Not applicable <!-- TaskRadio samples-pr -->
- A PR for the [documentation repository](https://github.com/radius-project/docs) is created, if the changes in this PR affect the documentation or any user facing updates are made.
    - [ ] Yes <!-- TaskRadio docs-pr -->
    - [x] Not applicable <!-- TaskRadio docs-pr -->
- A PR for the [recipes repository](https://github.com/radius-project/recipes) is created, if existing recipes are affected by the changes in this PR.
    - [ ] Yes <!-- TaskRadio recipes-pr -->
    - [x] Not applicable <!-- TaskRadio recipes-pr -->

ytimocin committed:
Air-gapped environment

Signed-off-by: ytimocin <ytimocin@microsoft.com>

ytimocin commented:
The gist that installs Radius on an air-gapped cluster: https://gist.github.com/ytimocin/8887d95ab1409562f4646fd30edb101c. **_Still WIP and needs to be updated._**

willdavsmith commented on pkg/cli/cmd/install/kubernetes/kubernetes.go:
```diff
@@ -76,6 +76,8 @@ rad install kubernetes --reinstall
 	commonflags.AddKubeContextFlagVar(cmd, &runner.KubeContext)
 	cmd.Flags().BoolVar(&runner.Reinstall, "reinstall", false, "Specify to force reinstallation of Radius")
 	cmd.Flags().StringVar(&runner.Chart, "chart", "", "Specify a file path to a helm chart to install Radius from")
+	cmd.Flags().StringVar(&runner.ContourChart, "contour-chart", "", "Specify a local file path to a helm chart to install Contour from")
```
do these have to be a local filepath or can they be an OCI registry (e.g. on a local CR)

ytimocin replied:
local path for now

willdavsmith commented on pkg/cli/helm/cluster.go:
```diff
@@ -94,6 +97,10 @@ func PopulateDefaultClusterOptions(cliOptions CLIClusterOptions) ClusterOptions
 		options.Radius.ChartPath = cliOptions.Radius.ChartPath
 	}
 
+	if cliOptions.Radius.ChartRepo != "" {
```
was this missing before?

ytimocin replied:
yes

willdavsmith commented on pkg/cli/helm/contourclient.go:
```diff
@@ -56,9 +59,30 @@ func ApplyContourHelmChart(options ContourOptions, kubeContext string) error {
-	if err != nil {
-		return fmt.Errorf("failed to get contour chart, err: %w, helm output: %s", err, helmOutput.String())
+	// Use provided repo URL if available, otherwise use default
+	repo := options.ChartRepo
```
can probably remove this codeblock, looks like dupe of below

willdavsmith reviewed (commented):
nice, I think this is the right place to add these changes. my only question is if we need to support loading a helm chart from a local container registry instead of a local file

nithyatsu commented:
Nice, LGTM. This will be a one off or would we officially support these options in main?

ytimocin committed:
Remove dapr flags

Signed-off-by: ytimocin <ytimocin@microsoft.com>

ytimocin committed:
Make Contour disabled by default

Signed-off-by: ytimocin <ytimocin@microsoft.com>

ytimocin committed:
Adding recipe-registry flag to the rad init command

Signed-off-by: ytimocin <ytimocin@microsoft.com>

ytimocin committed:
Adding recipeConfig.Terraform.Mirror and ProviderMappings etc.

Signed-off-by: ytimocin <ytimocin@microsoft.com>

ytimocin committed:
Merge branch 'main' into ytimocin/air-gapped-cluster

Signed-off-by: ytimocin <ytimocin@microsoft.com>

ytimocin committed:
Merge branch 'main' into ytimocin/air-gapped-cluster

Signed-off-by: ytimocin <ytimocin@microsoft.com>

ytimocin committed:
Registry changes for Terraform mirror

Signed-off-by: ytimocin <ytimocin@microsoft.com>

ytimocin committed:
Merge branch 'main' into ytimocin/air-gapped-cluster

ytimocin committed:
Terraform tls.skipVerify added to the terraform.version. We should talk about terraform.registry.authentication

Signed-off-by: ytimocin <ytimocin@microsoft.com>

ytimocin committed:
Merge branch 'main' into ytimocin/air-gapped-cluster

Signed-off-by: ytimocin <ytimocin@microsoft.com>

ytimocin committed:
Changing the terraform.version.releasesApiBaseURL logic

Signed-off-by: ytimocin <ytimocin@microsoft.com>

ytimocin committed:
Merge branch 'main' into ytimocin/air-gapped-cluster

ytimocin committed:
Add releasesArchiveURL for direct Terraform binary downloads

Signed-off-by: ytimocin <ytimocin@microsoft.com>

ytimocin committed:
TLS Configuration for Terraform Recipe Execution

Signed-off-by: ytimocin <ytimocin@microsoft.com>

ytimocin committed:
Merge branch 'main' into ytimocin/air-gapped-cluster

Signed-off-by: ytimocin <ytimocin@microsoft.com>

ytimocin committed:
Updating the env variables and the clients to handle the private registries

Signed-off-by: ytimocin <ytimocin@microsoft.com>

ytimocin committed:
Checkpoint: deploying a postgres works. There is a small issue with the credentials in the terraformrc file.

Signed-off-by: ytimocin <ytimocin@microsoft.com>

ytimocin committed:
Merge branch 'main' into ytimocin/air-gapped-cluster

ytimocin committed:
Fix Terraform recipe parameter serialization issues

- Handle empty maps/arrays passed as strings in recipe parameters
- Add recursive normalization for nested structures with string-encoded empty objects
- Support whitespace variations (e.g., '{ }', '{}\n', '[ ]')
- Remove TF_LOG environment variable to prevent terraform-exec conflicts
- Add comprehensive test coverage for parameter normalization

Signed-off-by: ytimocin <ytimocin@microsoft.com>

ytimocin committed:
Add updated.zip

Signed-off-by: ytimocin <ytimocin@microsoft.com>

sylvainsf committed:
Adds TF debug/trace logging support.  (#10042)

Test feature

Signed-off-by: Sylvain Niles <sylvainniles@microsoft.com>

sylvainsf committed:
Inject providers auth (#10107)

# Description

Bugfix and adds support for injecting azurerm credentials when provider
block is present.

## Type of change

<!--

Please select **one** of the following options that describes your
change and delete the others. Clearly identifying the type of change you
are making will help us review your PR faster, and is used in authoring
release notes.

If you are making a bug fix or functionality change to Radius and do not
have an associated issue link please create one now.

-->

- This pull request is a minor refactor, code cleanup, test improvement,
or other maintenance task and doesn't change the functionality of Radius
(issue link optional).


## Contributor checklist
Please verify that the PR meets the following requirements, where
applicable:

<!--
This checklist uses "TaskRadio" comments to make certain options
mutually exclusive.
See:
https://github.com/mheap/require-checklist-action?tab=readme-ov-file#radio-groups
For details on how this works and why it's required.
-->

- An overview of proposed schema changes is included in a linked GitHub
issue.
    - [ ] Yes <!-- TaskRadio schema -->
    - [x] Not applicable <!-- TaskRadio schema -->
- A design document PR is created in the [design-notes
repository](https://github.com/radius-project/design-notes/), if new
APIs are being introduced.
    - [ ] Yes <!-- TaskRadio design-pr -->
    - [x] Not applicable <!-- TaskRadio design-pr -->
- The design document has been reviewed and approved by Radius
maintainers/approvers.
    - [ ] Yes <!-- TaskRadio design-review -->
    - [x] Not applicable <!-- TaskRadio design-review -->
- A PR for the [samples
repository](https://github.com/radius-project/samples) is created, if
existing samples are affected by the changes in this PR.
    - [ ] Yes <!-- TaskRadio samples-pr -->
    - [x] Not applicable <!-- TaskRadio samples-pr -->
- A PR for the [documentation
repository](https://github.com/radius-project/docs) is created, if the
changes in this PR affect the documentation or any user facing updates
are made.
    - [ ] Yes <!-- TaskRadio docs-pr -->
    - [x] Not applicable <!-- TaskRadio docs-pr -->
- A PR for the [recipes
repository](https://github.com/radius-project/recipes) is created, if
existing recipes are affected by the changes in this PR.
    - [ ] Yes <!-- TaskRadio recipes-pr -->
    - [x] Not applicable <!-- TaskRadio recipes-pr -->

---------

Signed-off-by: Sylvain Niles <sylvainniles@microsoft.com>

sylvainsf committed:
sets the os env var for git binary to honor provided cert (#10181)

# Description

Sets the OS environment variable for git binary to use the provided
cert.

## Type of change

<!--

Please select **one** of the following options that describes your
change and delete the others. Clearly identifying the type of change you
are making will help us review your PR faster, and is used in authoring
release notes.

If you are making a bug fix or functionality change to Radius and do not
have an associated issue link please create one now.

-->

- This pull request is a minor refactor, code cleanup, test improvement,
or other maintenance task and doesn't change the functionality of Radius
(issue link optional).

<!--

Please update the following to link the associated issue. This is
required for some kinds of changes (see above).

-->



## Contributor checklist
Please verify that the PR meets the following requirements, where
applicable:

<!--
This checklist uses "TaskRadio" comments to make certain options
mutually exclusive.
See:
https://github.com/mheap/require-checklist-action?tab=readme-ov-file#radio-groups
For details on how this works and why it's required.
-->

- An overview of proposed schema changes is included in a linked GitHub
issue.
    - [ ] Yes <!-- TaskRadio schema -->
    - [x] Not applicable <!-- TaskRadio schema -->
- A design document PR is created in the [design-notes
repository](https://github.com/radius-project/design-notes/), if new
APIs are being introduced.
    - [ ] Yes <!-- TaskRadio design-pr -->
    - [x] Not applicable <!-- TaskRadio design-pr -->
- The design document has been reviewed and approved by Radius
maintainers/approvers.
    - [ ] Yes <!-- TaskRadio design-review -->
    - [x] Not applicable <!-- TaskRadio design-review -->
- A PR for the [samples
repository](https://github.com/radius-project/samples) is created, if
existing samples are affected by the changes in this PR.
    - [ ] Yes <!-- TaskRadio samples-pr -->
    - [x] Not applicable <!-- TaskRadio samples-pr -->
- A PR for the [documentation
repository](https://github.com/radius-project/docs) is created, if the
changes in this PR affect the documentation or any user facing updates
are made.
    - [ ] Yes <!-- TaskRadio docs-pr -->
    - [x] Not applicable <!-- TaskRadio docs-pr -->
- A PR for the [recipes
repository](https://github.com/radius-project/recipes) is created, if
existing recipes are affected by the changes in this PR.
    - [ ] Yes <!-- TaskRadio recipes-pr -->
    - [x] Not applicable <!-- TaskRadio recipes-pr -->

sylvainsf committed:
working url rewriting for terrafom.io calls

sylvainsf committed:
deprecate skip tls verify, move to network mirror as default.

sylvainsf committed:
deprecated the mirror type, fixed a bug in how it's added to terraformrc

sylvainsf committed:
recipe and providers all work

sylvainsf committed:
working build

sylvainsf committed:
remove microsft cert from testing

sylvainsf committed:
fix indent

sylvainsf committed:
indentation fix, removed unnecessary logic from terraformrc
